### DISCUSSION TITLE
Access private repos using Github App or OAuth App

### DISCUSSION THREAD

**Discussion by melchanti** [community/community#36463] [GitHub Education] [answered]
Hi,

I'm a student working to create a CICD pipeline using github app or Oauth App and github repos. I realized that most companies would have their repos private.

I looked through the documentation on Github App and OAuth App and I was able to figure out how to get a list of public repos but I'm unable to retrieve the private repos. I'm using Javascript to get the list of repos.

Below is my code, I tried the `/user/repos` with and without the `type` parameter, I also tried using the `visibility` parameter but no luck. Am I doing something wrong, is there another endpoint to retrieve the private repos that I may not be seeing?

```js
const octokit = new Octokit({
    auth: token
  });

  try {
    const response = await octokit.request('Get /user/repos?type=private', {});
    return response.data;
  } catch(e) {
    console.log(e.message);
  }
```

**Comment by melchanti** [comment, marked as answer]
After some digging and testing, I was able to figure out the problem. In order to view and access private repos, the Github App needs to be installed on the user's account. I was able to install my Github App on the account and from there I was able to access the private repos.

**Reply by iranzithierry** [reply]
how you done it

**Reply by abbasatayee** [reply]
each of the GitHub app has an installation url, using that you can install the app.

please follow the link.
https://docs.github.com/en/apps/using-github-apps

**Comment by ossanna16** [comment]
Hey there @melchanti and thank you for asking a good question! I'm glad you found our community 🙂

**Comment by abbasatayee** [comment]
The access token generated by a GitHub app will not work with the `/user/repos` endpoint. Instead, you should use the `/installation/repositories` endpoint to list all the repositories that the app has access to.

**Reply by abbasatayee** [reply]
This is a simple script that performs the task.

``` javascript
const fs = require('fs');
const jwt = require('jsonwebtoken');
const axios = require('axios');

// GitHub App installation ID
const installationId = 'INSTALATION_ID';

// GitHub App ID and path to the private key PEM file
const appId = 'APP_ID';
const privateKeyPath = 'KEY_PATH';

// Function to read the private key from PEM file
function getPrivateKey() {
    return fs.readFileSync(privateKeyPath, 'utf8');
}

// Function to fetch repositories based on installation ID
async function fetchRepositories() {
    try {
        // Generate JWT (JSON Web Token) for authentication
        const privateKey = getPrivateKey();
        const token = jwt.sign({ iss: appId }, privateKey, { algorithm: 'RS256', expiresIn: '10m' });

        // Get access token for the installation
        const accessTokenResponse = await axios.post(`https://api.github.com/app/installations/${installationId}/access_tokens`, {}, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Accept': 'application/vnd.github.v3+json'
            }
        });

        const accessToken = accessTokenResponse.data.token;
        console.log({accessToken})
        // Fetch repositories using the access token
        const reposResponse = await axios.get('https://api.github.com/installation/repositories', {
            headers: {
                'Authorization': `Bearer ${accessToken}`,
                'Accept': 'application/vnd.github.v3+json'
            }
        });
        // Log repository names
        const repositories = reposResponse.data;
        console.dir({repositories} , {depth : Infinity})
        // repositories.forEach(repo => {
        //     console.log(repo.full_name);
        // });

    } catch (error) {
        console.error('Error fetching repositories:', error.message);
    }
}

// Call the function to fetch repositories
fetchRepositories();
```

**Reply by iranzithierry** [reply]
can you filter by programming language here

**Reply by abbasatayee** [reply]
in which language you need the solution?

**Reply by denzuko** [reply]
@iranzithierry why, the code above is showing the bare minimum needed. One can filter in the app logic side by doing a three step process:

*    [Get an Installation Access Token](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-an-installation-access-token-for-a-github-app) from https://api.github.com/app/installations/{installation_id}/access_tokens, authenticating as the App (Bearer authorization, and a JWT signed by the App Private Key. The Installation Access Token expires 1 hour after issuance.

*    [Use the Installation Access Token](https://docs.github.com/en/rest/overview/authenticating-to-the-rest-api) as Bearer authorization token to query one of the [available API endpoints](https://docs.github.com/en/rest/overview/endpoints-available-for-github-app-installation-access-tokens), such as /installation/repositories. You won't be able to list the user's private repositories, but only those he granted your app access to.

*    Once you have an object of available repositories then one can iterate over that while using the [repository api](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28) to get the resulting repo objects and mapreduce the results based on the language/topics properties.
